Fining Revamp - getting scanned with illegal cargo gets you a new hail panel + more

quyykk opened the pull request:
## Feature Details

This PR introduces the following features:

- Fines now stack.
- If you're in space and a ship scanned you and found illegal cargo, you'll have several options like fleeing, paying the fine, or trying to bribe the ship.
- A new attribute for governments to specify the percentage of a fine it accepts as bribe (`"bribe factor"`)
- A new attribute for custom text when a governments finds that you have illegal stuff and if they can be bribed.

## UI Screenshots

![fine-revamp-illegal-cargo](https://user-images.githubusercontent.com/85879619/143778256-0495b6ab-7fc7-46f6-b8c3-ceb980bc389e.png)


## Usage Examples

```
government Gov
    "bribe factor" <val#>       # Percentage of fine that this gov wants as bribe. Defaults to 0.
    "interdiction" <msg>        # Custom message shown when this gov find illegal stuff in space.
    "interdiction bribe" <msg>  # Custom message shown if this gov can be bribed.
```

## Testing Done

Tested getting fined in space with various illegal outfits and cargo.

## Performance Impact

Negligible.

quyykk committed:
Fining revamp implementation.

quyykk committed:
Fix oversight that made it so that fines are always 0.

quyykk committed:
Fix fine stacking calculation.

quyykk committed:
Require "atrocity" to be positive.

Galaucus commented:
Is #5141 made redundant by this?

Ferociousfeind commented:
Quick thought(s)-
(also, poor merge conflicts)
any chance we could get the ability to surrender _some_ of our cargo, instead of all of it? Assuming that, when prompted, you would dump only the illegal stuff. I would imagine once you click "surrender cargo" you get the extra options- dump all, dump illegal, and dump some or similar.

I'm very excited for government-specific legalities getting into the game ASAP, it would open up lots of opportunities for strict regions or apathy in law.

quyykk commented:
> any chance we could get the ability to surrender some of our cargo, instead of all of it?

You're basically only surrendering your illegal cargo. I don't really see a reason for the player to choose to surrender everything though.

Ferociousfeind commented:
_shrug_
It may be standard practice to dump all your cargo. But, in that case, let me slightly amend my request-

Would it be possible to, when confronted by authorities, give the player the option to refuse surrender, or surrender all their illegal cargo, or surrender only a portion of their illegal cargo? A way to still make a buck, if the coppers don't realize you aren't being honest. There's a chance they didn't pick up some of that cargo, and betraying your whole hand is costly. It might be worth it to risk hostilities for a little extra profit. A game of limited informationa.

quyykk commented:
It's already possible to refuse to surrender, in which case you provoke the gov and they start attacking you. For the other request, idk maybe.

petervdmeer commented on source/Government.cpp:
```diff
@@ -116,6 +118,8 @@ void Government::Load(const DataNode &node)
 			bribe = child.Value(1);
 		else if(child.Token(0) == "fine" && child.Size() >= 2)
 			fine = child.Value(1);
+		else if(child.Token(0) == "bribe factor" && child.Size() >= 2)
```
I would expect to also see a `"reputation hit"` factor when the player flees. (Not a large factor for most governments, but fleeing should have a little bit longer-term effects.)

petervdmeer commented:
> Would it be possible to, when confronted by authorities, give the player the option to refuse surrender, or surrender all their illegal cargo, or surrender only a portion of their illegal cargo? A way to still make a buck, if the coppers don't realize you aren't being honest. There's a chance they didn't pick up some of that cargo, and betraying your whole hand is costly. It might be worth it to risk hostilities for a little extra profit. A game of limited information.

For this PR I would suggest to just focus on surrendering all illegal cargo and to keep those ideas for a possible follow-up PR. (I would assume the authorities in most cases to board the offending ship and take the illegal cargo directly btw).

petervdmeer commented:
"Waiting on OP" due to the open merge-conflicts.

Galaucus commented:
I prefer an all or nothing approach when it comes to surrendering cargo. It keeps things simple and straightforward.

petervdmeer commented:
A related PR contains a [nice comment](https://github.com/endless-sky/endless-sky/pull/2650#issuecomment-528461178) about allow fines to both have a "once once discovered" and a "per item when discovered" fine part. I feel that such an approach would also be nice here.

tibetiroka commented on source/IllegalHailPanel.cpp:
```diff
@@ -0,0 +1,133 @@
+
+Endless Sky is distributed in the hope that it will be useful, but WITHOUT ANY
+WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+PARTICULAR PURPOSE.  See the GNU General Public License for more details.
```
```suggestion
PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with
this program. If not, see <https://www.gnu.org/licenses/>.
```
The standard copyright header was changed recently.

tibetiroka commented on source/IllegalHailPanel.h:
```diff
@@ -0,0 +1,58 @@
+
+Endless Sky is distributed in the hope that it will be useful, but WITHOUT ANY
+WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+PARTICULAR PURPOSE.  See the GNU General Public License for more details.
```
```suggestion
PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with
this program. If not, see <https://www.gnu.org/licenses/>.
```

tibetiroka commented on source/PlayerHailPanel.cpp:
```diff
@@ -0,0 +1,289 @@
+
+Endless Sky is distributed in the hope that it will be useful, but WITHOUT ANY
+WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+PARTICULAR PURPOSE.  See the GNU General Public License for more details.
```
```suggestion
PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with
this program. If not, see <https://www.gnu.org/licenses/>.
```

tibetiroka commented on source/PlayerHailPanel.h:
```diff
@@ -0,0 +1,71 @@
+
+Endless Sky is distributed in the hope that it will be useful, but WITHOUT ANY
+WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+PARTICULAR PURPOSE.  See the GNU General Public License for more details.
```
```suggestion
PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with
this program. If not, see <https://www.gnu.org/licenses/>.
```

Amazinite commented:
@quyykk Please address the conflicts and outstanding reviews on this PR, or we'll have to close it.

quyykk committed:
Apply suggestions from code review

Co-authored-by: tibetiroka <68112292+tibetiroka@users.noreply.github.com>

quyykk committed:
Merge remote-tracking branch 'upstream/master' into fining-revamp

quyykk committed:
Add missing files to CMakeLists.

quyykk committed:
Add missing parantheses.

quyykk committed:
Fix merge issue.

quyykk committed:
Fix style issues.

Hurleveur commented:
The illegal tests need to be changed a bit to account for the new panel, idk what changes happened to it.

quyykk committed:
Merge remote-tracking branch 'upstream/master' into fining-revamp

quyykk committed:
Fix style violations.

quyykk committed:
Fix the integration tests.

Hurleveur commented on source/CargoHold.cpp:
```diff
@@ -602,15 +601,15 @@ int CargoHold::IllegalCargoFine(const Government *government) const
 		if(fine < 0)
 			return fine;
-		totalFine = max(totalFine, fine / 2);
+		totalFine += (fine / 2) * it.second;
```
Ok to make the fines stack but maybe use a sqrt or smth?
Or maybe we should just rebalancr fines afterwards, but 450M caus of nerve gas is going to hurt (which is good), possibly too much

quyykk replied:
Yeah definitely

Hurleveur commented on source/Government.cpp:
```diff
@@ -216,6 +218,25 @@ void Government::Load(const DataNode &node)
+				seen = true;
+			}
+
+			const auto &value = child.Token(1);
```
```suggestion
			const string &value = child.Token(1);
```
Usually we just get the string dont we?

Hurleveur commented on source/Government.cpp:
```diff
@@ -216,6 +218,25 @@ void Government::Load(const DataNode &node)
+			const auto &value = child.Token(1);
+			if(!text.empty() && !value.empty() && value[0] > ' ')
+				text += '\t';
+			text += child.Token(1);
```
```suggestion
			text += value;
```

Hurleveur commented on source/Government.cpp:
```diff
@@ -231,6 +252,8 @@ void Government::Load(const DataNode &node)
 			bribe = add ? bribe + child.Value(valueIndex) : child.Value(valueIndex);
 		else if(key == "fine")
 			fine = add ? fine + child.Value(valueIndex) : child.Value(valueIndex);
+		else if(child.Token(0) == "bribe factor" && child.Size() >= 2)
```
Use key instead of child.token(0)
Also child.size >=2 is the same as hasvalue I think, so no need to check it (line243)
```suggestion
		else if(key == "bribe factor")
```

Hurleveur commented on source/Government.cpp:
```diff
@@ -216,6 +218,25 @@ void Government::Load(const DataNode &node)
 		else if(key == "provoked on scan")
 			provokedOnScan = true;
+		else if((child.Token(0) == "interdiction" || child.Token(0) == "interdiction bribe")
+				&& child.Size() >= 2)
```
Use key instead of child.tokem(0) and move this after the value check of line 243 so we dont have to check child.size

Hurleveur reviewed (approved):
Just looked at the parsing for now

Hurleveur commented:
Phone github... I didnt click aprove xD

samrocketman committed:
Resolved conflics: Merge branch 'master'

Update warp-core changes for HailPanel in PlayerHailPanel (cpp and h) for #8360

Updated changes from #8236 across moved code for `Format::CreditString`.
- IllegalHailPanel.cpp
- PlayerHailPanel.cpp

samrocketman committed:
Merge branch 'master' into resolve-conflicts-fining-revamp

quyykk committed:
Merge remote-tracking branch 'upstream/master' into fining-revamp

quyykk committed:
Fix merge conflict

quyykk committed:
Address review comments

quyykk committed:
Merge remote-tracking branch 'upstream/master' into fining-revamp

quyykk committed:
Fix unused capture.

Hurleveur commented on source/IllegalHailPanel.cpp:
```diff
@@ -0,0 +1,138 @@
+
+	static const std::string defaultMessage
+		= "You've been detected carrying illegal <type> and have been issued a fine of <fine>."
+			" \n\tDump your cargo immediately or we'll be forced to disable and board your ship.";
```
```suggestion
			" \n\tDump your cargo immediately or we'll be forced to open fire.";
```
seems better as a default to me?

Hurleveur commented on source/IllegalHailPanel.cpp:
```diff
@@ -0,0 +1,138 @@
+		// is automatically removed since the missions are failed.
+		for(const auto &pair : scannedShip.Cargo().Outfits())
+			if(pair.first->Get("illegal") || pair.first->Get("atrocity") > 0.)
+				scannedShip.Jettison(pair.first, pair.second, true);
```
say you have illegal cargo, but also atrocities, but they've only discovered one
should you be forced to dump all of it?
or is it always discover all/nothing?

Hurleveur commented on source/IllegalHailPanel.cpp:
```diff
@@ -0,0 +1,138 @@
+
+		player.Accounts().AddCredits(-bribe);
+		Messages::Add("You bribed a " + hailingShip.GetGovernment()->GetName() + " ship "
+			+ Format::CreditString(bribe) + " to avoid paying a fine today."
```
I just find it weird you can bribe them to avoid the fine lol, one would just take the lowest cost option without much consequences? also more generic caus it could be to ignore the atrocity (or not?)
```suggestion
			+ Format::CreditString(bribe) + " to leave you alone."
```

Hurleveur commented on source/PlayerHailPanel.cpp:
```diff
@@ -0,0 +1,326 @@
+	else
+		header = ship->DisplayModelName() + " (" + gov->GetName() + "): ";
+	// Drones are always unpiloted, so they never respond to hails.
+	bool isMute = ship->GetPersonality().IsMute() || (ship->Attributes().Category() == "Drone");
```
now, ik you didnt write this but I think we should change it eventually given the new CanSendHail method and the possibility for aliens with unknown languages to say smth more interesting than you dont recognize stuff

Hurleveur commented on source/PlayerHailPanel.cpp:
```diff
@@ -0,0 +1,326 @@
+	else if(gov->IsEnemy())
+	{
+		// Enemy ships always show hostile messages.
+		// They either show bribing messages,
```
that's on me ig but that's not the case anymore they can show normal msgs now instead of the bribing all the time

Hurleveur commented on source/Politics.cpp:
```diff
@@ -237,98 +237,76 @@ bool Politics::HasDominated(const Planet *planet) const
+
+	// Do nothing if there is nothing to fine for or the target ship has already been fined today.
+	if(!punishment.HasPunishment() || fined.count(ship.GetGovernment()))
+		return {};
```
shouldnt we have a const static empty for this?

Hurleveur reviewed (changes requested):
Some things to consider

I hope this can be the beginning of a more dynamic hail system for npcs where they can eventually give you missions and the like

anyway I didnt look at all of it but this should give you some work already, code looks good and I'll be testing this at some point to make sure all's working fine

tibetiroka commented:
Closing as orphaned as there are unresolved comments from over a year ago, and several merge conflicts with master. Feel free to reopen this PR once these are addressed.
